src/optimization/DAG.cpp

```diff
@@ -0,0 +1,149 @@
+//
+// Created by nomaddo on 17/12/18.
+//
+
+#include <log.h>
+#include <assert.h>
+#include <iostream>
+#include <functional>
+#include <algorithm>
+#include <string>
+#include <stack>
+#include <utility>
+#include <vector>
+
+#include "DAG.h"
+#include "../InstructionWalker.h"
+#include "log.h"
+
+#include "../ControlFlowGraph.h"
+#include "../DebugGraph.h"
+#include "../InstructionWalker.h"
+#include "../Profiler.h"
+
+namespace vc4c {
+
+/* XXX should be replaced with common routines that generate such instructions? */
+/* reading from mutex I/O register means getting mutex */
+static bool isGetLocInstruction (IL * instr) {
+	for (Value v: instr->getArguments()) {
+		if (v.hasRegister(REG_MUTEX)) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+/* writing to mutex I/O register means release of mutex */
+static bool isReleaseLocInstruction (IL * instr) {
+	if (instr->getOutput().has_value()) {
+		auto output = instr->getOutput().value();
+		return output.hasRegister(REG_MUTEX);
+	}
+
+	return false;
+}
+
+DAG::DAG(BasicBlock &bb) {
+	std::map<Value, IL*> map;
+  auto mutexInstructions = std::shared_ptr<std::vector<IL*>>(new std::vector<IL*>());
+	IL * getMutexInstr = nullptr;
+
+	auto setFlagInstructions = std::shared_ptr<std::vector<IL*>>(new std::vector<IL*>());
+	IL * setFlagInstr = nullptr;
+
+	for (auto it = bb.begin(); !it.isEndOfBlock(); it.nextInBlock()) {
+		auto instr = it.get();
+
+		/* Instruction scheduler is responsible for translation to CombinedOperation.
+		 * Not allowed to translate before!
+		 */
+		assert(! dynamic_cast<vc4c::intermediate::CombinedOperation*>(instr));
```

Comment:
As mentioned above, this can be asserted by moving the scheduler-pass before the combiner-pass.